core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -1300,24 +1298,31 @@ public NameResolverRegistry getNameResolverRegistry() {
     public void onResult(final ResolutionResult resolutionResult) {
       final class NamesResolved implements Runnable {
 
-        @SuppressWarnings("ReferenceEquality")
+        @SuppressWarnings({"ReferenceEquality", "unchecked"})
         @Override
         public void run() {
           List<EquivalentAddressGroup> servers = resolutionResult.getAddresses();
+          if (servers.isEmpty() && lastResolutionResultWasSuccess == Boolean.TRUE) {
```

Comment:
According to the "Address handling" section from the spreadsheet, empty address should be applied to LB if LB doesn't need address. So, config should be ignored but the empty addresses should be passed to `tryHandleResolvedAddresses()`. To make the decision we need to know whether LB needs addresses, thus it has to be done inside `tryHandleResolvedAddresses()` with `waitingForServiceConfig` passed to it too.

Above is about the addresses. For config, it'd be better to make the decision in the main if-tree below. It's easier to follow if all decisions are made in the same place.

Reply:
resolved PTAL